integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -136,11 +136,13 @@ public fun <T> ListenableFuture<T>.asDeferred(): Deferred<T> {
         override fun onSuccess(result: T?) {
             // Here we work with flexible types, so we unchecked cast to trick the type system
             @Suppress("UNCHECKED_CAST")
-            deferred.complete(result as T)
+            runCatching { deferred.complete(result as T) }
+                .onFailure { handleCoroutineException(EmptyCoroutineContext, it) }
         }
 
         override fun onFailure(t: Throwable) {
-            deferred.completeExceptionally(t)
+            runCatching { deferred.completeExceptionally(t) }
+                .onFailure { handleCoroutineException(EmptyCoroutineContext, it) }
```

Comment:
Not sure if it's intentional, but this is inconsistent with the way we decided to handle undeliverable exceptions in the reactive integrations: https://github.com/Kotlin/kotlinx.coroutines/blob/master/reactive/kotlinx-coroutines-reactor/src/Flux.kt#L58-L67 The discrepancy here is that the suppressed exception, which could potentially be useful, is not being set.

The same is true about the other implementations.

Reply:
I believe that this a different use case.
Ones in RX is `use`-like, where we fail during (potentially user-supplied) exception handling on cancellation.

This is a completely different story, we're delivering the exception (not handling it, just doing the regular processing) and something unexpected and/or off-the-track happened: internal errors which we have no other means to deliver or exception from `invokeOnCancellation`, which is basically the same. Here this exception is the dominant one, not the one that was going to be deferred result